presto-jmx/src/main/java/com/facebook/presto/connector/jmx/JmxHistoryDumper.java

```diff
@@ -0,0 +1,106 @@
+
+public class JmxHistoryDumper
+        implements Runnable
+{
+    private static final Logger log = Logger.get(JmxHistoryDumper.class);
+
+    private final JmxHistoricalData jmxHistoricalData;
+    private final JmxMetadata jmxMetadata;
+    private final JmxRecordSetProvider jmxRecordSetProvider;
+    private final ScheduledExecutorService executor = newSingleThreadScheduledExecutor();
+    private final long period;
+    private long lastDumpTimestamp;
+
+    @Inject
+    public JmxHistoryDumper(
+            JmxHistoricalData jmxHistoricalData,
+            JmxMetadata jmxMetadata,
+            JmxRecordSetProvider jmxRecordSetProvider,
+            JmxConnectorConfig jmxConfig)
+    {
+        this.jmxHistoricalData = requireNonNull(jmxHistoricalData, "jmxStatsHolder is null");
+        this.jmxMetadata = requireNonNull(jmxMetadata, "jmxMetadata is null");
+        this.jmxRecordSetProvider = requireNonNull(jmxRecordSetProvider, "jmxRecordSetProvider is null");
+        this.period = requireNonNull(jmxConfig, "jmxConfig is null").getDumpPeriod().toMillis();
+    }
+
+    public void start()
+    {
+        long nowMillis = getNowMillis();
+        long initialDelay = nowMillis - roundToPeriod(nowMillis + period);
+        executor.scheduleAtFixedRate(this, initialDelay, period, MILLISECONDS);
+    }
+
+    private long roundToPeriod(long millis)
+    {
+        return ((millis + period / 2) / period) * period;
+    }
+
+    @Override
+    public void run()
+    {
+        // we are using rounded up timestamp, so that records from different nodes and different
+        // tables will have matching timestamps (for joining/grouping etc)
+        long dumpTimestamp = roundToPeriod(getNowMillis());
+
+        // if something has lagged and next dump has the same timestamp as this one, ignore this dump
+        if (dumpTimestamp == lastDumpTimestamp) {
+            return;
+        }
+        lastDumpTimestamp = dumpTimestamp;
+
+        for (String tableName : jmxHistoricalData.getTables()) {
+            try {
+            JmxTableHandle tableHandle = requireNonNull(
+                    jmxMetadata.getTableHandle(new SchemaTableName(JmxMetadata.HISTORY_SCHEMA_NAME, tableName)),
+                    format("tableHandle is null for table [%s]", tableName));
+
+                List<Object> row = jmxRecordSetProvider.getLiveRow(
+                        tableHandle,
+                        jmxRecordSetProvider.getColumnTypes(tableHandle.getColumnHandles()),
+                        dumpTimestamp);
+                jmxHistoricalData.addRow(tableName, row);
+            }
+            catch (Throwable ex) {
+                log.error(ex, "Error starting JmxHistoryDumper");
+            }
+        }
+    }
+
+    public void shutdown()
+    {
+        executor.shutdown();
+    }
+
+    private static long getNowMillis()
+    {
+        return MILLISECONDS.convert(System.nanoTime(), NANOSECONDS);
```

Comment:
Ok, I was using `currentTimeMillis` but later someone pointed out this: https://github.com/prestodb/presto/pull/3737 so I switched to `nanoTime`. I guess that we will have to  live with this monotonic issue of `currentTimeMillis` - I have adjusted sanity check for ignoring dump when:
```
        long dumpTimestamp = roundToPeriod(currentTimeMillis());

        // if something has lagged and next dump has the same timestamp as this one, or somehow
        // time on the machine was changed backward  (monotonic issue of currentTimeMillis)
        // ignore this dump
        if (dumpTimestamp <= lastDumpTimestamp) {
            return;
        }
```